File: packages/health_connector_hc_android/android/src/test/kotlin/com/phamtunglam/health_connector_hc_android/unit_tests/mappers/health_measurement_unit_mappers/MassMapperTest.kt
package com.phamtunglam.health_connector_hc_android.unit_tests.mappers.health_measurement_unit_mappers

import androidx.health.connect.client.units.Mass
import com.phamtunglam.health_connector_hc_android.mappers.health_measurement_unit_mappers.toDto
import com.phamtunglam.health_connector_hc_android.mappers.health_measurement_unit_mappers.toHealthConnect
import com.phamtunglam.health_connector_hc_android.pigeon.MassDto
import io.kotest.matchers.shouldBe
import org.junit.jupiter.api.DisplayName
import org.junit.jupiter.api.Test
import org.junit.jupiter.api.TestInstance

/**
 * Unit tests for Mass Mapper.
 *
 * Tests verify proper bidirectional mapping between [MassDto] and Health Connect
 * [Mass] objects, including unit conversions between kilograms, grams, pounds, and ounces.
 */
@DisplayName("MassMapper")
@TestInstance(TestInstance.Lifecycle.PER_CLASS)
class MassMapperTest {

    private companion object {
        const val TEST_VALUE = 100.0
    }

    @Test
    @DisplayName(
        "GIVEN MassDto in kilograms → " +
            "WHEN toHealthConnect called → " +
            "THEN creates Mass with correct value",
    )
    fun whenMassDtoInKilograms_thenCreatesCorrectMass() {
        // Given
        val dto = MassDto(kilograms = TEST_VALUE)

        // When
        val result = dto.toHealthConnect()

        // Then
        result.inKilograms shouldBe TEST_VALUE
    }

    @Test
    @DisplayName(
        "GIVEN Mass object → " +
            "WHEN toDto called → " +
            "THEN converts to DTO in kilograms",
    )
    fun whenMassToDto_thenUsesKilograms() {
        // Given
        val mass = Mass.kilograms(TEST_VALUE)

        // When
        val result = mass.toDto()

        // Then
        result.kilograms shouldBe TEST_VALUE
    }
}

File: packages/health_connector_hc_android/test/unit_tests/src/mappers/measurement_unit_mappers/mass_mapper_test.dart
import 'package:health_connector_core/health_connector_core_internal.dart';
import 'package:health_connector_hc_android/src/mappers/measurement_unit_mappers/mass_mapper.dart';
import 'package:health_connector_hc_android/src/pigeon/health_connector_hc_android_api.g.dart';

import 'package:test/test.dart';

void main() {
  group(
    'MassMapper',
    () {
      group(
        'MassToDto',
        () {
          test(
            'converts Mass to MassDto in kilograms',
            () {
              const mass = Mass.kilograms(70.0);
              final dto = mass.toDto();

              expect(dto.kilograms, 70.0);
            },
          );

          test(
            'converts Mass from pounds to kilograms',
            () {
              const mass = Mass.pounds(154.324);
              final dto = mass.toDto();

              expect(dto.kilograms, closeTo(70.0, 0.1));
            },
          );
        },
      );

      group(
        'MassDtoToDomain',
        () {
          test(
            'maps MassDto to Mass',
            () {
              final dto = MassDto(kilograms: 70.0);
              final mass = dto.toDomain();

              expect(mass.inKilograms, closeTo(70.0, 0.0001));
            },
          );
        },
      );
    },
  );
}
